File: snake.py
import random 
from inputimeout import inputimeout, TimeoutOccurred

width = 20
height = 20

class Cords:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y

class Snake:
    snakeParts: list[Cords] = []
    score = len(snakeParts)
    def __init__(self):
        self.snakeParts.append(Cords(10,10))
    def moveY(self, dir):
        # end game if it touches itself 
        self.snakeParts.insert(0, Cords(self.snakeParts[0].x, self.snakeParts[0].y - dir))
        print(f"{self.snakeParts[0].x} {self.snakeParts[0].y}")
        if (self.snakeParts[0].x >= width or self.snakeParts[0].x < 0) or (self.snakeParts[0].y >= height or self.snakeParts[0].y < 0): 
            # if out of bounds, game should end
            print("GAME OVER\nSnake out of bounds")
            exit()

        if (self.snakeParts[0].y in (yv.y for yv in self.snakeParts[1:])): 
            # checks if snake is touching itself by comaring the head value to the read of the values 
            print("GAME OVER\nSnake cannot touch itself") # seems to glitch and counts being next to as an end
            exit()
        if self.snakeParts[0].x == location.x and self.snakeParts[0].y == location.y:
            screen.newApple() # if its touching the apple is creates a new apple
        else: 
            self.snakeParts.pop() # if it isnt it removes the last part to create new motion
            # should remove the last element to make it appear like its moving forward
    def moveX(self, dir):
        self.snakeParts.insert(0, Cords(self.snakeParts[0].x + dir, self.snakeParts[0].y))
        print(f"{self.snakeParts[0].x} {self.snakeParts[0].y}")
        if (self.snakeParts[0].x >= width or self.snakeParts[0].x < 0) or (self.snakeParts[0].y >= height or self.snakeParts[0].y < 0):
            print("GAME OVER\nSnake out of bounds")
            exit()

        if (self.snakeParts[0].x in (xv.x for xv in self.snakeParts[1:])): 
            # checks if snake is touching itself by comaring the head value to the read of the values 
            print("GAME OVER\nSnake cannot touch itself") # seems to glitch and counts being next to as an end
            exit()
        if self.snakeParts[0].x == location.x and self.snakeParts[0].y == location.y:
            screen.newApple() # if its touching the apple is creates a new apple
        else: 
            self.snakeParts.pop() # if it isnt it removes the last part to create new motion
            # should remove the last element to make it appear like its moving forward
 
class Screen:
    screenElements: list = []
    def __init__(self):
        for i in range(width*height):
            self.screenElements.append(' ')

    def newApple(self): # make into a class itself maybe
        spot = random.randint(0,width*height)
        global location
        location = Cords(spot%width, spot//height)
        # save location 

        if self.screenElements[spot] == ' ': # to ensure it spawns on empty space
            self.screenElements[spot] = "\uf8ff"
        else:
            self.newApple()

    def clearScreen(self):
        for i in range(len(self.screenElements)):
            if self.screenElements[i] == "\uf8ff":
                self.screenElements[i] = "\uf8ff"
            else:
                self.screenElements[i] = ' '

    def updScreen(self, snake: Snake):
        for part in snake.snakeParts: 
            self.screenElements[part.y*width + part.x] = "\u25A0"

    def printScreen(self): 
        i = 1
        for element in self.screenElements:
            print(element, end=' ')
            if i % width == 0:
                print(end="\n")
            i+=1

def init():
    global screen
    global userSnake
    screen = Screen()
    userSnake = Snake()
    screen.newApple()
    screen.updScreen(userSnake)
    screen.printScreen()

def update():
    screen.clearScreen()
    screen.updScreen(userSnake)
    screen.printScreen()

init()

last_dir = 'w'
while(True):
    try: 
        dir = inputimeout(prompt="wasd: ", timeout=1)
    except TimeoutOccurred:
        dir = last_dir
    match dir:
        case 'w': 
            userSnake.moveY(1) 
        case 's':
            userSnake.moveY(-1) 
        case 'a':
            userSnake.moveX(-1)
        case 'd':
            userSnake.moveX(1)
        case _:
            print("input not found")
    last_dir = dir
    update()

File: snake.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Linq.Expressions;
using System.Security.Cryptography.X509Certificates;

// currently does not end game if you make self contact
// needs to run without input

namespace Snake
{
    static class Globals
    {
        public const int WIDTH = 20;
        public const int HEIGHT = 20;
        public static Random random = new Random();
    }
    class Cords
    {
        public int x;
        public int y;
        public Cords(int x1, int y1)
        {
            x = x1;
            y = y1;
        }
    }
    class Snake
    {
        public List<Cords> snakeSegs = new List<Cords>(); 
        public int score => snakeSegs.Count;
        private Cords start = new Cords(10,10);
        public Snake() {
            snakeSegs.Add(start);
        }
        public void movey(int dir)
        {
            snakeSegs.Insert(0, new Cords(snakeSegs[0].x, snakeSegs[0].y - dir)); 
            if (snakeSegs[0].x >= Globals.WIDTH || snakeSegs[0].x < 0 || 
            snakeSegs[0].y >= Globals.HEIGHT || snakeSegs[0].y < 0)
            {
                Console.WriteLine("out of bounds, game over");
                Console.WriteLine("Score: " + score);
                Environment.Exit(0);
            }
            if (snakeSegs[0].x == Progwram.screen.location.x && snakeSegs[0].y == Program.screen.location.y)
            {
                Program.screen.newApple();

            } else
            {
                snakeSegs.RemoveAt(snakeSegs.Count - 1);
            }
        }
        public void movex(int dir)
        {
            snakeSegs.Insert(0, new Cords(snakeSegs[0].x + dir, snakeSegs[0].y)); 
            if (snakeSegs[0].x >= Globals.WIDTH || snakeSegs[0].x < 0 || 
            snakeSegs[0].y >= Globals.HEIGHT || snakeSegs[0].y < 0)
            {
                Console.WriteLine("out of bounds, game over");
                Console.WriteLine("Score: " + score);
                Environment.Exit(0);
            }
            if (snakeSegs[0].x == Program.screen.location.x && snakeSegs[0].y == Program.screen.location.y)
            {
                Program.screen.newApple();
            } else
            {
                snakeSegs.RemoveAt(snakeSegs.Count - 1);
            }
        }
    }
    class Screen
    {
        public Cords location = null!;
        public List<string> elements = new List<string>();
        public Screen()
        {
            for(int i = 1; i <= Globals.WIDTH*Globals.HEIGHT; i++)
            {
                elements.Add(" ");
            }
        }

        public void newApple()
        {
            int spot = Globals.random.Next(Globals.WIDTH*Globals.HEIGHT);
            location = new Cords(spot%Globals.WIDTH, spot / Globals.WIDTH);
            if (elements[spot] == " ")
            {
                elements[spot] = "\uf8ff";
            } else
            {
                newApple();
            }
        }

        public void clearScreen()
        {
            for(int i = 0; i < elements.Count; i++)
            {
                if (elements[i] == "\uf8ff")
                {
                    elements[i] = "\uf8ff";
                } else
                {
                    elements[i] = " ";
                }
            }
        }

        public void updateScreen(Snake snake)
        {
            elements[location.y * Globals.WIDTH + location.x] = "\uf8ff";
            foreach(Cords part in snake.snakeSegs)
            {
                elements[part.y*Globals.WIDTH + part.x] = "\u25A0"; 
            }
        }

        public void printScreen()
        {
            int i = 1;
            foreach(string part in elements)
            {
                Console.Write(part + " ");
                if (i % Globals.WIDTH == 0)
                {
                    Console.WriteLine();
                }
                i += 1;
            }
        }
    }
    class Program
    {
        public static Screen screen = null!;
        public static Snake userSnake = null!;
        public static void init()
        {
            screen = new Screen();
            userSnake = new Snake();
            screen.newApple();
            screen.updateScreen(userSnake);
            screen.printScreen();
        }
        public static void update()
        {
            screen.clearScreen();
            screen.updateScreen(userSnake);
            screen.printScreen();
        }
        static void Main()
        {
            init();
            while (true) {
                Console.Write("wasd: ");
                string? input = Console.ReadLine();

                if (string.IsNullOrEmpty(input))
                    continue;

                char dir = input[0];              
                switch (dir)
                {
                    case 'w' : userSnake.movey(1); break;
                    case 's' : userSnake.movey(-1); break;
                    case 'a' : userSnake.movex(-1); break;
                    case 'd' : userSnake.movex(1); break;
                    default : Console.WriteLine("Imput not known"); break;
                }
                update();
            }
        }
    }
}